src/v/storage/storage_resources.h

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/property.h"
+#include "units.h"
+
+#include <seastar/core/semaphore.hh>
+
+#include <cstdint>
+
+namespace storage {
+
+/**
+ * This class is extension of ss::semaphore to fit the needs
+ * of the storage_resources class's tracking of byte/concurrency
+ * allowances.
+ *
+ * Callers may use this class as either a soft or hard quota.  In
+ * the hard case, regular async-waiting semaphore calls (ss::get_units)
+ * may be used.  In the soft case, the take() function will allow the
+ * semaphore count to go negative, but return a `checkpoint_hint` field
+ * that prompts the holder of the units to release some.
+ *
+ * This is 'adjustable' in that:
+ * - Regular sempahores are just a counter: they have no
+ *   memory of their intended capacity.  In order to enable runtime
+ *   changes to the max units in a semaphore, we must keep an extra
+ *   record of the capacity.
+ * - This enables runtime configuration changes to parameters that
+ *   control the capacity of a semaphore.
+ */
+class adjustable_allowance : public ss::semaphore {
```

Comment:
how much of the semaphore interface is used to justify inheritance here? for example, we have `mutex` which embeds a semaphore. but I don't have a strong argument for either case.

Reply:
I pondered that while writing it.  I ended up only using a couple of actual semaphore methods so it's probably not justified (I prefer encapsulation too).  I added a commit that refactors this.